### ISSUE TITLE
PinSage sampler stochastic despite setting seeds 

### ISSUE THREAD

**Issue by Rsalganik1123** [dmlc/dgl#5099]
## 🐛 Bug

I want to ensure that my sampler returns the same neighbourhood for a set of input nodes each time the code is triggered. Currently, setting the seed doesn't guarantee that PinSage sampler shows stochastic behaviour. That is, each time I run this code, it will return different neighbourhoods for the same input nodes, even if I use the same seed. 

I am doing hyperparameter tuning using a dispatch system but I can't compare between the performance of runs because it's so dependent on the neighbours being selected (this is both for training and inference). 
 
## To Reproduce
```
import dgl
import torch 
import sys 
import pickle
import numpy as np 
import random 

#Fixing all seeds 
dgl.seed(47)
np.random.seed(47)
torch.manual_seed(47)
torch.cuda.manual_seed_all(47)
torch.cuda.manual_seed(47)   
random.seed(47)
torch.backends.cudnn.enabled=False
torch.backends.cudnn.deterministic=True


#Initialize graph 
num_playlists = 110
num_tracks = 10_000
num_edges = 12_000
src = torch.randint(0, 100, (num_edges, ))
dst = torch.randint(0, 10_000, (num_edges, ))
data_dict = {
     ('playlist', 'contains', 'track'): (src, dst),
     ('track', 'contained_by', 'playlist'): (dst, src),
 }
g = dgl.heterograph(data_dict)

#Generate seed nodes 
seeds = torch.tensor([4188, 1259, 5869, 5113,  391, 9271,  702, 2867, 6661, 8017, 2573,  578,
         751, 3074, 3552])

#Build Sampler 
def compact_and_copy(frontier, seeds):
    block = dgl.to_block(frontier, seeds)
    for col, data in frontier.edata.items():
        if col == dgl.EID:
            continue
        block.edata[col] = data[block.edata[dgl.EID]]
    return block

class NeighborSampler(object):
    def __init__(self, g, user_type, item_type, random_walk_length, random_walk_restart_prob,
                 num_random_walks, num_neighbors):
        self.g = g
        self.seed = 47
        self.user_type = user_type
        self.item_type = item_type
        self.user_to_item_etype = list(g.metagraph()[user_type][item_type])[0]
        self.item_to_user_etype = list(g.metagraph()[item_type][user_type])[0]
        self.sampler = dgl.sampling.PinSAGESampler(g, item_type, user_type, random_walk_length,
                                        random_walk_restart_prob, num_random_walks, num_neighbors)
            

    def sample_blocks(self, seeds, heads=None, tails=None, neg_tails=None, remove_self_loops=False ):
        blocks = []
        # dgl.seed(self.seed)
        frontier = self.sampler(seeds)
        if heads is not None:
            eids = frontier.edge_ids(torch.cat([heads, heads]), torch.cat([tails, neg_tails]), return_uv=True)[2]
            if len(eids) > 0:
                old_frontier = frontier
                frontier = dgl.remove_edges(old_frontier, eids)
        block = compact_and_copy(frontier, seeds)
        seeds = block.srcdata[dgl.NID]
        blocks.insert(0, block)
        return blocks

sampler1 = NeighborSampler(g, 'playlist', 'track', 2, 0.5, 10, 2)
b1 = sampler1.sample_blocks(seeds)
print("RUN {}:{}".format(sys.argv[-1], b1[0].ndata[dgl.NID]['track'])) 
# pickle.dump(b1, open('./b{}.pkl'.format(sys.argv[-1]), "wb"))
```
This produces an output that seems to oscillate between two versions: 
```
RUN 1:tensor([4188, 1259, 5869, 5113,  391, 9271,  702, 2867, 6661, 8017, 2573,  578,
         751, 3074, 3552, 3824, 9687, 9401, 9002, 3434, 8508, 9477, 9293, 3734,
        9481, 8619, 9660, 7915, 8455, 9675, 8702, 8183, 7977, 9955, 9829, 9933,
        9369, 9916, 7646, 6604, 9768])

RUN 2:tensor([4188, 1259, 5869, 5113,  391, 9271,  702, 2867, 6661, 8017, 2573,  578,
         751, 3074, 3552, 7509, 7862, 1337, 9461, 9995, 9720, 9200, 9022, 8714,
        8189,  118, 8358, 8148, 8138, 7825, 1484, 9732, 8846, 9861, 9708, 8611,
        7735, 8466, 1990, 5961, 8639])

RUN 3:tensor([4188, 1259, 5869, 5113,  391, 9271,  702, 2867, 6661, 8017, 2573,  578,
         751, 3074, 3552, 3824, 9687, 9401, 9002, 3434, 8508, 9477, 9293, 3734,
        9481, 8619, 9660, 7915, 8455, 9675, 8702, 8183, 7977, 9955, 9829, 9933,
        9369, 9916, 7646, 6604, 9768])

RUN 4:tensor([4188, 1259, 5869, 5113,  391, 9271,  702, 2867, 6661, 8017, 2573,  578,
         751, 3074, 3552, 3824, 9687, 9401, 9002, 3434, 8508, 9477, 9293, 3734,
        9481, 8619, 9660, 7915, 8455, 9675, 8702, 8183, 7977, 9955, 9829, 9933,
        9369, 9916, 7646, 6604, 9768])

RUN 5:tensor([4188, 1259, 5869, 5113,  391, 9271,  702, 2867, 6661, 8017, 2573,  578,
         751, 3074, 3552, 7509, 7862, 1337, 9461, 9995, 9720, 9200, 9022, 8714,
        8189,  118, 8358, 8148, 8138, 7825, 1484, 9732, 8846, 9861, 9708, 8611,
        7735, 8466, 1990, 5961, 8639])

```

## Expected behavior

Since the seed is set, the output should be exactly the same each run, that is, runs 1 - 5 should have the same output: 
```
RUN X:tensor([4188, 1259, 5869, 5113,  391, 9271,  702, 2867, 6661, 8017, 2573,  578,
         751, 3074, 3552, 7509, 7862, 1337, 9461, 9995, 9720, 9200, 9022, 8714,
        8189,  118, 8358, 8148, 8138, 7825, 1484, 9732, 8846, 9861, 9708, 8611,
        7735, 8466, 1990, 5961, 8639])
```

## Environment

 - DGL Version (e.g., 1.0): 8.2
 - Backend Library & Version (e.g., PyTorch 0.4.1, MXNet/Gluon 1.3): PyTorch 1.11.0
 - OS (e.g., Linux): Linux
 - How you installed DGL (`conda`, `pip`, source): Honda 
 - Build command you used (if compiling from source):
 - Python version: 3.9.3
 - CUDA/cuDNN version (if applicable): 11.1
 - GPU models and configuration (e.g. V100): V100
